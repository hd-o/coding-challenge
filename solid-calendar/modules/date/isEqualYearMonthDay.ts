import { createContext } from 'react'

function matchYearMonthDate (a: Date, b: Date): boolean {
  return (
    a.getUTCFullYear() === b.getUTCFullYear() &&
    a.getUTCMonth() === b.getUTCMonth() &&
    a.getUTCDate() === b.getUTCDate()
  )
}

export const DateMatchYearMonthDate = createContext(matchYearMonthDate)
