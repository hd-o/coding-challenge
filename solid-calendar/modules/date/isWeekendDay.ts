import { createContext, useContext } from 'react'
import { LodashMemoizeCtx } from '~/pkg/lodash/memoize'

let dateIsWeekendDay: (index: number) => boolean

function useDateIsWeekendDay (): typeof dateIsWeekendDay {
  const memoize = useContext(LodashMemoizeCtx)
  return dateIsWeekendDay ?? (dateIsWeekendDay = (
    memoize((index: number) => (index - 6) % 7 === 0 || index % 7 === 0)
  ))
}

export const DateIsWeekendDayCtx = createContext(useDateIsWeekendDay)
