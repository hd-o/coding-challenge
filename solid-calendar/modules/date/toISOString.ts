import { createContext, useContext } from 'react'
import { DateFnsFormatISO9075Ctx } from '~/pkg/date-fns/formatISO9075'

export function useDateToISOString (): (date: Date) => string {
  const formatISO9075 = useContext(DateFnsFormatISO9075Ctx)
  return (date) => formatISO9075(date).replace(' ', 'T')
}

export const DateToISOStringCtx = createContext(useDateToISOString)
