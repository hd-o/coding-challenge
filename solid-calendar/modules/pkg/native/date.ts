import { createContext } from 'react'

export const NativeDateCtx = createContext(Date)
