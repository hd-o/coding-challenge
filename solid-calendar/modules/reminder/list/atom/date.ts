import { WritableAtom } from 'jotai'
import { createContext, useContext } from 'react'
import { JotaiAtomCtx } from '~/pkg/jotai/atom'

type AtomValue = Date | null

let reminderListDateAtom: WritableAtom<AtomValue, AtomValue>

function useReminderListDateAtom (): typeof reminderListDateAtom {
  const atom = useContext(JotaiAtomCtx)
  return reminderListDateAtom ?? (reminderListDateAtom = atom<AtomValue>(null))
}

export const ReminderListDateAtomCtx = createContext(useReminderListDateAtom)
