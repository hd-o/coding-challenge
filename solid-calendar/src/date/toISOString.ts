import { DateFnsFormatISO9075Ctx } from '/src/pkg/date-fns/formatISO9075'
import { createContext, useContext } from 'react'

export function useDateToISOString (): (date: Date) => string {
  const formatISO9075 = useContext(DateFnsFormatISO9075Ctx)
  return (date) => formatISO9075(date).replace(' ', 'T')
}

export const DateToISOStringCtx = createContext(useDateToISOString)
