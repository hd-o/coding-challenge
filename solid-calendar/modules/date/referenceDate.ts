import { createContext, useContext } from 'react'
import { NativeDateCtx } from '~/pkg/native/date'

let referenceDate: Date

/**
 * Used by date utils, like date-fns parse
 * @see {@link https://date-fns.org/docs/parse#syntax|date-fns/parse}
 */
function useReferenceDate (): Date {
  const Date = useContext(NativeDateCtx)
  return referenceDate ?? (referenceDate = new Date())
}

export const ReferenceDateCtx = createContext(useReferenceDate)
