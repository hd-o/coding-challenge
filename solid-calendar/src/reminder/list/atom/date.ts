import { JotaiAtomCtx } from '/src/pkg/jotai/atom'
import { CacheCtx } from '/src/util/cache'
import { WritableAtom } from 'jotai'
import { createContext, useContext } from 'react'

type AtomType = Date | null

function useReminderListDateAtom (): WritableAtom<AtomType, AtomType> {
  const atom = useContext(JotaiAtomCtx)
  return useContext(CacheCtx)(
    'reminderListDateAtom', [atom], () => atom<AtomType>(null))
}

export const ReminderListDateAtomCtx = createContext(useReminderListDateAtom)
