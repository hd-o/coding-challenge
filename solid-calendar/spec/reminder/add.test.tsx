import { DateToISOStringCtx } from '/src/date/toISOString'
import { subMonths } from 'date-fns'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { date } from '../calendar/util/date'
import { renderCalendar } from '../calendar/util/render'
import { renderAddReminderReminderEditor } from '../calendar/util/render/addReminderReminderEditor'
import { fireChange } from '../util/fire/change'
import { getContext } from '../util/get/context'
import { getReminderEditorState } from './util/get/reminderEditorState'
import { renderAddedReminder } from './util/render/addedReminder'
import { renderReminderEditorLocation } from './util/render/reminderEditorLocation'
import { renderReminderList } from './util/render/reminderList'
import { renderReminderListAddReminder } from './util/render/reminderListAddReminder'

beforeEach(() => {
  renderCalendar()
})

describe('calendar date "add reminder" button', () => {
  test('opens reminder editor when clicked', async () => {
    await renderAddReminderReminderEditor()
    expect(screen.getByTestId('reminder-editor')).toBeTruthy()
  })
})

describe('reminder editor', () => {
  beforeEach(async () => {
    await renderAddReminderReminderEditor()
  })
  test('reminder editor is correctly filled when opened', async () => {
    expect(await getReminderEditorState()).toMatchSnapshot('editor-open-data')
  })
  test('location search, and weather', async () => {
    await renderReminderEditorLocation()
    expect(await getReminderEditorState()).toMatchSnapshot('editor-location')
    // Shows location weather info
    const weather = screen.getByTestId('reminder-editor-weather')
    expect(weather.textContent).toMatchSnapshot('editor-location-weather')
    // Hides weather info on location clear
    userEvent.click(screen.getByTitle('Clear'))
    expect(screen.queryByTestId('reminder-editor-weather')).toBeNull()
  })
  test('toggles location weather unit', async () => {
    await renderReminderEditorLocation()
    const weather = screen.getByTestId('reminder-editor-weather')
    expect(weather.textContent).toMatchSnapshot('editor-location-weather-unit')
    userEvent.click(weather)
    expect(weather.textContent).toMatchSnapshot('editor-location-weather-unit-change')
  })
  test('prevents save when title not filled', async () => {
    const saveBtn = screen.getByText('Save Reminder')
    expect(saveBtn.getAttribute('disabled')).not.toBeNull()
  })
  test('closes, and clears data on cancel', async () => {
    userEvent.type(screen.getByTestId('reminder-editor-title'), 'Test')
    expect(await getReminderEditorState()).toMatchSnapshot('editor-cancel-title-input')
    userEvent.click(screen.getByText('Cancel'))
    expect(screen.queryByTestId('reminder-editor')).toBeNull()
    await renderAddReminderReminderEditor()
    expect(await getReminderEditorState()).toMatchSnapshot('editor-cancel-reopen')
  })
  test('adds reminder on save', async () => {
    await renderAddedReminder()
    userEvent.click(screen.getByTestId('calendar-cell-reminder-count'))
    expect(screen.getByTestId('reminder-list-item').textContent).toMatchSnapshot('editor-reminder-saved')
  })
})

describe('reminder list "add reminder" button', () => {
  beforeEach(async () => {
    await renderAddReminderReminderEditor()
    await renderAddedReminder()
    await renderReminderList()
  })
  test('added reminder is shown on list', async () => {
    expect(screen.getByTestId('reminder-list-item').textContent).toMatchSnapshot('list-add')
  })
  test('added reminder with other date is not shown on list', async () => {
    await renderReminderListAddReminder()
    // Insert date of previous month
    const toISOString = getContext(DateToISOStringCtx)
    const dateInput: HTMLInputElement = screen.getByTestId('reminder-editor-date')
    fireChange(dateInput, toISOString(subMonths(date, 1)))
    // Save reminder, and compare snapshot
    await renderAddedReminder()
    expect(screen.getByTestId('reminder-list-item').textContent).toMatchSnapshot('list-add-other-date')
  })
})
