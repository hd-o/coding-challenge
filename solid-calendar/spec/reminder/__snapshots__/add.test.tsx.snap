// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reminder editor adds reminder on save: editor-reminder-saved 1`] = `"TestWed Dec 01 2021 01:00:00 GMT-0300 (Brasilia Standard Time)Barcelona • 🌦️ Showers 23°C / 23°C"`;

exports[`reminder editor closes, and clears data on cancel: editor-cancel-reopen 1`] = `
Object {
  "date": "12/01/2021 01:00 am",
  "location": "",
  "title": "",
  "weather": "",
}
`;

exports[`reminder editor closes, and clears data on cancel: editor-cancel-title-input 1`] = `
Object {
  "date": "12/01/2021 01:00 am",
  "location": "",
  "title": "Test",
  "weather": "",
}
`;

exports[`reminder editor location search, and weather: editor-location 1`] = `
Object {
  "date": "12/01/2021 01:00 am",
  "location": "Barcelona",
  "title": "",
  "weather": "Barcelona • 🌦️ Showers 23°C / 23°C",
}
`;

exports[`reminder editor location search, and weather: editor-location-weather 1`] = `"Barcelona • 🌦️ Showers 23°C / 23°C"`;

exports[`reminder editor reminder editor is correctly filled when opened: editor-open-data 1`] = `
Object {
  "date": "12/01/2021 01:00 am",
  "location": "",
  "title": "",
  "weather": "",
}
`;

exports[`reminder editor toggles location weather unit: editor-location-weather-unit 1`] = `"Barcelona • 🌦️ Showers 23°C / 23°C"`;

exports[`reminder editor toggles location weather unit: editor-location-weather-unit-change 1`] = `"Barcelona • 🌦️ Showers 23°C / 23°C"`;

exports[`reminder list "add reminder" button added reminder is shown on list: list-add 1`] = `"TestWed Dec 01 2021 01:00:00 GMT-0300 (Brasilia Standard Time)Barcelona • 🌦️ Showers 23°C / 23°C"`;

exports[`reminder list "add reminder" button added reminder with other date is not shown on list: list-add-other-date 1`] = `"TestWed Dec 01 2021 01:00:00 GMT-0300 (Brasilia Standard Time)Barcelona • 🌦️ Showers 23°C / 23°C"`;
